## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1065]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [1066]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [1067]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [1068]:
# What three item categories had the most entries?
top3_cat = df["category"].value_counts().head(3)
top3_cat
# consumables, furniture, and software had the most entries.

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [1069]:
# For the category with the most entries,
# which subcategory had the most entries?
consumables_df = df[df["category"] == "consumables"]
consumables_df
consumables_top_subcat = consumables_df['subcategory'].value_counts()
consumables_top_subcat
# bathroom supplies had the most entries in the consumables category.

subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
misc                 2891
pens                 2603
Name: count, dtype: int64

In [1070]:
# Which five clients had the most entries in the data?
df['full_name'] = df['first'] + ' ' + df['last']
df.drop(columns=['first', 'last'], inplace = True)
new_order = ['full_name'] + [col for col in df.columns if col != 'full_name']
df = df[new_order]
df
client_counts = df.groupby(['full_name', 'client_id']).size().sort_values(ascending=False)
client_counts

full_name        client_id
Jessica Reyes    33615        220
Angela Everett   66037        211
Bryan Myers      46820        209
Kendra Garrett   24741        207
Alexandra Young  38378        207
                             ... 
Alicia Butler    70734          4
Angela Morales   32001          4
Karen Chase      77583          3
Emily Anthony    76017          3
Kelly Watson     94987          1
Length: 999, dtype: int64

In [1071]:
# Use this space to do any additional research
# and familiarize yourself with the data.



In [1072]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = client_counts.head(5)
top_5_clients

full_name        client_id
Jessica Reyes    33615        220
Angela Everett   66037        211
Bryan Myers      46820        209
Kendra Garrett   24741        207
Alexandra Young  38378        207
dtype: int64

In [1073]:
# How many total units (the qty column) did the
# client with the most entries order order?
top_client_JR = df[df['full_name'] == 'Jessica Reyes']
top_client_qty = top_client_JR['qty'].sum()
top_client_qty


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [1074]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df


full_name                      job               phone  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  
0                1  115164.00  
1                0     523.95  
2                6     527.28  
3                3    1056.18  
4                1    3902.00  
...            ...        ...  
54634            8    2743.29  
54635            0    9709.73  
54636            6   31188.50  
54637            0     165.76  
54638            5    3103.31  

[54639 rows x 18 columns]

In [1075]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
def calculate_shipping_price(row):
    total_weight = row['qty'] * row['unit_weight']
    if total_weight > 50:
        return 7 * total_weight
    else:
        return 10 * total_weight

df['shipping_price'] = df.apply(calculate_shipping_price, axis=1)
df


full_name                      job               phone  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  shipping_price  
0                1  115164.00         5512.50  
1                0     523.95          312.90  
2                6     527.28          458.64  
3                3    1056.18          356.70  
4                1    3902.00         6500.20  
...            ...        ...             ...  
54634            8    2743.29          519.75  
54635            0    9709.73         3849.30  
54636            6   31188.50        13832.00  
54637            0     165.76        14143.36  
54638            5    3103.31        14939.19  

[54

In [1076]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
def calculate_total_price(row):
    subtotal = row['subtotal']
    shipping_price = row['shipping_price']
    sales_tax = 0.0925 * (subtotal + shipping_price)
    total_price = subtotal + shipping_price + sales_tax
    return total_price

df['total_price'] = df.apply(calculate_total_price, axis=1)
df 



full_name                      job               phone  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number   subtotal  shipping_price    total_price  
0                1  115164.00         5512.50  131839.076250  
1                0     523.95          312.90     914.258625  
2                6     527.28          458.64    1077.117600  
3                3    1056.18          356.70    1543.571400  
4                1    3902.00         6500.20   11364.403500  
...            ...        ...             ...            ...  
54634            8    2743.29          519.75    3564.871200  
54635            0    9709.73         3849.30   14813.240275  
54636        

In [1077]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
def calculate_line_cost (row):
    unit_cost = row['unit_cost']
    qty = row['qty']
    shipping_price = row['shipping_price']
    line_cost = (unit_cost * qty) + shipping_price
    return line_cost

df['line_cost'] = df.apply(calculate_line_cost, axis=1)
df


full_name                      job               phone  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id  ...         subcategory  \
0              17        2023  EUD29711-63-6U  ...            wall art   
1              20        2023  XDA18116-89-4A  ...                pens   
2               4        2023  ABE59463-05-7E  ...  project management   
3              17        2023  ZMM00836-65-0C  ...                pens   
4              21        2023  BZX55559-12-3X  ...                misc   
...           ...         ...             ...  ...                 ...   
54634          17        2023  UCA28532-31-0B  ...           logistics   
54635           6        2023  EBD29277-23-7U  ...    kitchen supplies   
54636           9        2023  EBD30093-55-9Z  ...   bathroom supplies   
54637           3        2023  XMZ38239-49-3D  ...              stands   
54638           1        2023  ZCZ05589-50-5D  ...  project management   

      unit_price  unit_cost  unit_weight   qty  line_number   subtotal  \
0        1096.80     762.71         7.50   105            1  115164.00   
1          24.95      15.09         1.49    21            0     523.95   
2          13.52       7.86         1.68    39            6     527.28   
3          36.42      24.85         1.23    29            3    1056.18   
4         195.10     108.17        46.43    20            1    3902.00   
...          ...        ...          ...   ...          ...        ...   
54634      83.13      51.60         2.25    33            8    2743.29   
54635     206.59     175.46        11.70    47            0    9709.73   
54636      65.66      57.31         4.16   475            6   31188.50   
54637       1.48       1.22        18.04   112            0     165.76   
54638       3.01       1.61         2.07  1031            5    3103.31   

       shipping_price    total_price  line_cost  
0             5512.50  131839.076250   85597.05  
1              312.90     914.258625     629.79  
2              458.64    1077.117600     765.18  
3              356.70    1543.571400    1077.35  
4             6500.20   11364.403500    8663.60  
...               ...            ...        ...  
54634          519.75    3564.871200    2222.55  


In [1078]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
df


full_name                      job               phone  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id  ... unit_price unit_cost  \
0              17        2023  EUD29711-63-6U  ...    1096.80    762.71   
1              20        2023  XDA18116-89-4A  ...      24.95     15.09   
2               4        2023  ABE59463-05-7E  ...      13.52      7.86   
3              17        2023  ZMM00836-65-0C  ...      36.42     24.85   
4              21        2023  BZX55559-12-3X  ...     195.10    108.17   
...           ...         ...             ...  ...        ...       ...   
54634          17        2023  UCA28532-31-0B  ...      83.13     51.60   
54635           6        2023  EBD29277-23-7U  ...     206.59    175.46   
54636           9        2023  EBD30093-55-9Z  ...      65.66     57.31   
54637           3        2023  XMZ38239-49-3D  ...       1.48      1.22   
54638           1        2023  ZCZ05589-50-5D  ...       3.01      1.61   

       unit_weight   qty  line_number   subtotal  shipping_price  \
0             7.50   105            1  115164.00         5512.50   
1             1.49    21            0     523.95          312.90   
2             1.68    39            6     527.28          458.64   
3             1.23    29            3    1056.18          356.70   
4            46.43    20            1    3902.00         6500.20   
...            ...   ...          ...        ...             ...   
54634         2.25    33            8    2743.29          519.75   
54635        11.70    47            0    9709.73         3849.30   
54636         4.16   475            6   31188.50        13832.00   
54637        18.04   112            0     165.76        14143.36   
54638         2.07  1031            5    3103.31        14939.19   

         total_price  line_cost        profit  
0      131839.076250   85597.05  46242.026250  
1         914.258625     629.79    284.468625  
2        1077.117600     765.18    311.937600  
3        1543.571400    1077.35    466.221400  
4       11364.403500    8663.60   2700.803500  
...              ...        ...           ...  
54634    3564.871200    2222.55   1342.321200  
54635   14813.240275   12095.92   2717.320275  
54636   49184.896250   41054

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [1079]:
order_2742071 = df.loc[df['order_id'] == 2742071, 'total_price'].sum()
order_2173913 = df.loc[df['order_id'] == 2173913, 'total_price'].sum()
order_6128929 = df.loc[df['order_id'] == 6128929, 'total_price'].sum()
print('Order ID 2742071 had a total price of: ', order_2742071)
print('Order ID 2173913 had a total price of: ', order_2173913)
print('Order ID 6128929 had a total price of: ', order_6128929)


Order ID 2742071 had a total price of:  152811.897075
Order ID 2173913 had a total price of:  162388.71929999997
Order ID 6128929 had a total price of:  923441.24045


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [1080]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
client_total_spend = df.groupby('full_name')['total_price'].sum()
top_5_spenders = client_total_spend.nlargest(5)
top_5_spenders


full_name
Julie Anderson      6.559237e+08
Dennis Hawkins      2.374170e+08
Michele Mcdonald    1.835526e+08
Amy Stewart         1.812663e+08
Lauren Jones        1.066481e+08
Name: total_price, dtype: float64

In [1081]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
client_summary = df.groupby('full_name').agg({'qty': 'sum', 'shipping_price': 'sum', 'total_price': 'sum', 'profit': 'sum'})

top_5_clients_summary = client_summary.nlargest(5, 'profit')

summary_df = pd.DataFrame({'Total Units Purchased': top_5_clients_summary['qty'], 'Total Shipping Price': top_5_clients_summary['shipping_price'], 'Total Revenue': top_5_clients_summary['total_price'], 'Total Profit': top_5_clients_summary['profit']})

summary_df = summary_df.sort_values(by='Total Profit', ascending=False)
summary_df


Total Units Purchased  Total Shipping Price  Total Revenue  \
full_name                                                                      
Julie Anderson                  4033302            5340756.88   6.559237e+08   
Dennis Hawkins                   644423           10122549.19   2.374170e+08   
Michele Mcdonald                 402770            4391003.73   1.835526e+08   
Amy Stewart                     1010426           64335727.72   1.812663e+08   
Kendra Garrett                   239862            5126448.37   8.226889e+07   

                  Total Profit  
full_name                       
Julie Anderson    2.473528e+08  
Dennis Hawkins    1.092941e+08  
Michele Mcdonald  4.335173e+07  
Amy Stewart       4.297470e+07  
Kendra Garrett    3.657999e+07

In [1082]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
rename_df = df.rename(columns={'full_name': 'Full Name', 'job': 'Job', 'phone': 'Phone Number', 'email': 'Email Address', 'client_id': 'Client ID', 'order_id': 'Order ID', 'order_date': 'Order Date', 'order_week': 'Order Week', 'order_year': 'Order Year', 'item_id': 'Item ID', 'category': 'Category', 'subcategory': 'Subcategory', 'unit_price': 'Unit Price', 'unit_cost': 'Unit Cost', 'unit_weight': 'Unit Weight', 'qty': 'Quantity Ordered', 'line_number': 'Line Number', 'subtotal': 'Subtotal', 'shipping_price': 'Shipping Price', 'total_price': 'Total Price', 'line_cost': 'Line Cost', 'profit': 'Profit'})
df = rename_df
df

def format_currency_to_millions(x):
    return '${:,.5f}M'.format(x / 10**6)

columns_to_format = ['Unit Price', 'Unit Cost', 'Subtotal', 'Shipping Price', 'Total Price', 'Line Cost', 'Profit']
df[columns_to_format] = df[columns_to_format].map(format_currency_to_millions)
df

Full Name                      Job        Phone Number  \
0          Donald Harding             Immunologist  793-904-7725x39308   
1           Tiffany Myers          Music therapist    201.442.4543x942   
2          Shannon Watson             Immunologist   687.737.9424x8503   
3            Nathan Baker    Accounting technician    827-788-8123x012   
4      Christina Schwartz             Chiropractor        265-829-3643   
...                   ...                      ...                 ...   
54634        Lauren Reese           Radio producer  977.877.5272x11382   
54635       Derrick Moore                 Musician        358.661.5483   
54636    Monica Gutierrez         Graphic designer    294.805.9100x339   
54637       Wanda Solomon             Toxicologist       (311)767-4924   
54638         Dwayne Bell  Chief Financial Officer   606-619-7667x9508   

                             Email Address  Client ID  Order ID  Order Date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       Order Week  Order Year         Item ID  ... Unit Price  Unit Cost  \
0              17        2023  EUD29711-63-6U  ...  $0.00110M  $0.00076M   
1              20        2023  XDA18116-89-4A  ...  $0.00002M  $0.00002M   
2               4        2023  ABE59463-05-7E  ...  $0.00001M  $0.00001M   
3              17        2023  ZMM00836-65-0C  ...  $0.00004M  $0.00002M   
4              21        2023  BZX55559-12-3X  ...  $0.00020M  $0.00011M   
...           ...         ...             ...  ...        ...        ...   
54634          17        2023  UCA28532-31-0B  ...  $0.00008M  $0.00005M   
54635           6        2023  EBD29277-23-7U  ...  $0.00021M  $0.00018M   
54636           9        2023  EBD30093-55-9Z  ...  $0.00007M  $0.00006M   
54637           3        2023  XMZ38239-49-3D  ...  $0.00000M  $0.00000M   
54638           1        2023  ZCZ05589-50-5D  ...  $0.00000M  $0.00000M   

      Unit Weight Quantity Ordered  Line Number   Subtotal  Shipping Price  \
0            7.50              105            1  $0.11516M       $0.00551M   
1            1.49               21            0  $0.00052M       $0.00031M   
2            1.68               39            6  $0.00053M       $0.00046M   
3            1.23               29            3  $0.00106M       $0.00036M   
4           46.43               20            1  $0.00390M       $0.00650M   
...           ...              ...          ...        ...             ...   
54634        2.25               33            8  $0.00274M       $0.00052M   
54635       11.70               47            0  $0.00971M       $0.00385M   
54636        4.16              475            6  $0.03119M       $0.01383M   
54637       18.04              112            0  $0.00017M       $0.01414M   
54638        2.07             1031            5  $0.00310M       $0.01494M   

      Total Price  Line Cost     Profit  
0       $0.13184M  $0.08560M  $0.04624M  
1       $0.00091M  $0.00063M  $0.00028M  
2       $0.00108M  $0.00077M  $0.00031M  
3       $0.00154M  $0.00108M  $0.00047M  
4       $0.01136M  $0.00866M  $0.00270M  
...           ...        ...        ...  
54634   $0.00356M  $0.00222M  $0.0

In [1083]:
# Sort the updated data by "Total Profit" form highest to lowest
df_sorted = df.sort_values(by='Profit', ascending=False)
df_sorted

Full Name                                       Job  \
40161  Charles Mccormick  Conservation officer, historic buildings   
11444       James Brooks                        Charity fundraiser   
29737     William Savage                         Minerals surveyor   
23532       Susan Glover               Industrial/product designer   
26518      Taylor Barker                              Tree surgeon   
...                  ...                                       ...   
34222     Barbara Hester                           Camera operator   
7167         Heather Day                Regulatory affairs officer   
30396   Katherine Hardin                       Magazine journalist   
33731   Sarah Richardson       Exhibitions officer, museum/gallery   
11409  Christopher Hobbs               Health and safety inspector   

                Phone Number                      Email Address  Client ID  \
40161  001-667-790-6949x2885  mccormick_sign@potter-sanford.com      46571   
11444   001-866-994-6097x017            brooks.j@morgan-fox.com      18789   
29737           534.689.6772          savagewilliam@johnson.com      30339   
23532   001-874-645-1061x557                  sglover@ellis.com      20849   
26518       970.537.6082x931          t.barker@jones-ibarra.net      38974   
...                      ...                                ...        ...   
34222     446-648-4635x25731          hester.site.3268@king.com      41491   
7167       (251)730-3328x301                    dayh@hughes.net      43607   
30396           229-575-6791             hardink5344@lewis.info      78465   
33731             2056834802    richardsonsmile4695@merritt.com      79361   
11409           663-527-5878              hobbsc9635@weaver.com      34507   

       Order ID  Order Date  Order Week  Order Year         Item ID  ...  \
40161   1553623  2023-05-22          21        2023  EMC36732-26-6B  ...   
11444   5037924  2023-05-29          22        2023  MDU78510-11-2X  ...   
29737   9375270  2023-02-14           7        2023  BDC65078-60-7M  ...   
23532   7126774  2023-01-08           1        2023  CUE23412-94-3X  ...   
26518   3566343  2023-01-18           3        2023  ZBZ76765-83-2X  ...   
...         ...         ...         ...         ...             ...  ...   
34222   5694121  2023-04-08          14        2023  CMZ60736-39-9D  ...   
7167    9569726  2023-05-22          21        2023  ADE36146-76-0A  ...   
30396   7520308  2023-02-06           6        2023  DAC03696-47-4B  ...   
33731   3370363  2023-04-19          16        2023  EUD29711-63-6U  ...   
11409   4486634  2023-01-26           4        2023  DMB07325-21-0B  ...   

      Unit Price  Unit Cost Unit Weight Quantity Ordered  Line Number  \
40161  $0.00010M  $0.00006M       23.49           127184            6   
11444  $0.00041M  $0.00038M        2.51           109328            0   
29737  $0.00023M  $0.00020M       27.51           100479            7   
23532  $0.00039M  $0.00025M        0.12            39337            2   
26518  $0.00022M  $0.00016M        1.33            75729            5   
...          ...        ...         ...              ...          ...   
34222  $0.00001M  $0.00001M        2.70                0            2   
7167   $0.00012M  $0.00007M        2.30                0            1   
30396  $0.00017M  $0.00016M        2.62                0            6   
33731  $0.00110M  $0.00076M        7.50                0            4   
11409  $0.00000M  $0.00000M        0.28               11            1   

         Subtotal  Shipping Price Total Price   Line Cost     Profit  
40161  $12.64845M      $20.91287M  $36.66574M  $28.55281M  $8.11293M  
11444  $44.49759M       $1.92089M  $50.71219M  $43.36386M  $7.34833M  
29737  $23.08405M      $19.34924M  $46.35837M  $39.28227M  $7.07610M  
23532  $15.35953M       $0.03304M  $16.81638M   $9.98609M  $6.83029M  
26518  $16.81411M       $0.70504M  $19.13967M  $12.48620M  $6.65347M  
...           ...    